meson.build

```diff
@@ -0,0 +1,266 @@
+  default_options: [
+    'c_std=gnu99',
+    'warning_level=2',
+    'buildtype=release',
+    'libdir=/usr/lib64',
+    'prefix=/usr']
+  )
+
+# get feature options
+with_security = get_option('security').enabled()
+with_slp = get_option('slp').enabled()
+enable_shared = get_option('shared').enabled()
+enable_shared_version = get_option('shared_version').enabled()
+enable_static = get_option('static').enabled()
+
+if not enable_shared and not enable_static
+  error('You must enable either shared or static libary building')
+endif
+
+# get directory name override options
+man_dir = get_option('mandir')
+etc_dir = get_option('sysconfdir')
+systemd_unit_dir = get_option('systemddir') / 'system'
+sbin_dir = get_option('sbindir')
+run_dir = get_option('rundir')
+
+subdir('tests')
+subdir('doc')
+subdir('etc')
+subdir('include/libisns')
+
+lib_srcs = files([
+  'server.c',
+  'client.c',
+  'objects.c',
+  'callback.c',
+  'timer.c',
+  'vendor.c',
+  'db.c',
+  'db-file.c',
+  'db-policy.c',
+  'relation.c',
+  'scope.c',
+  'message.c',
+  'security.c',
+  'authblock.c',
+  'policy.c',
+  'pki.c',
+  'register.c',
+  'query.c',
+  'getnext.c',
+  'deregister.c',
+  'esi.c',
+  'scn.c',
+  'dd.c',
+  'entity.c',
+  'portal-group.c',
+  'storage-node.c',
+  'domain.c',
+  'simple.c',
+  'tags.c',
+  'attrs.c',
+  'export.c',
+  'socket.c',
+  'slp.c',
+  'error.c',
+  'logging.c',
+  'config.c',
+  'parser.c',
+  'buffer.c',
+  'pidfile.c',
+  'sysdep-unix.c',
+  'util.c',
+  'bitvector.c',
+  'mdebug.c'])
+
+incl_dirs = include_directories(['.', 'include'])
+
+security_dep = []
+if with_security
+  security_dep = dependency('libcrypto')
+endif
+
+add_project_arguments(
+  ['-D_GNU_SOURCE'],
+  language: 'c')
+
+# we will need this below
+cc = meson.get_compiler('c')
+
+#
+# build our config.h file
+#
+
+conf = configuration_data()
+conf.set('WITH_SECURITY', with_security)
+conf.set('WITH_SLP', with_slp)
+conf.set('HAVE_SLP_H', with_slp)
+conf.set('RUNDIR', run_dir)
+
+config_h = configure_file(
+  output: 'config.h',
+  configuration: conf)
+
+config_incdir = include_directories('.')
+
+config_dep = declare_dependency(
+  include_directories: config_incdir,
+  sources: config_h)
+
+#
+# build our library -- we always build either the static or
+# the shared library, or both
+#
+
+lib_deps = [config_dep, security_dep]
+if with_slp
+  lib_deps += cc.find_library('slp')
+endif
+
+if enable_static
+  libisns_lib = static_library('isns',
+    lib_srcs,
+    include_directories: incl_dirs,
+    dependencies: lib_deps,
+    install: true)
+endif
+
+if enable_shared
+  if enable_shared_version
+    shared_link_args = '-Wl,--version-script=@0@/libisns.vers'.format(meson.current_source_dir())
+  else
+    shared_link_args = []
+  endif
+  libisns_lib = shared_library('isns',
+    lib_srcs,
+    include_directories: incl_dirs,
+    dependencies: lib_deps,
+    link_args: shared_link_args,
+    version: '0',
+    install: true)
+endif
```

Comment:
I think you can just use `libisns_lib = library('isns', ...)`

It accepts the kwargs of both static and shared libraries, and can build one or both, depending on the value of `-Ddefault_library={static|shared|both}`. It will just use whichever arguments are relevant to whichever type it builds.

This is built into Meson, the same way --enable-shared and --enable-static is built into libtool.

Reply:
That seems much better -- except for the --version-script part. Is there something I can pass that will work all the time, i.e. meson figures out if it's okay or not? I looked for some 'version_script' argument to library() but didn't see one.

Reply:
It's just the same argument you'd pass to shared_library -- i.e. link_args. Those will be ignored if you never link, and static libraries don't link.

It's a goal of mine to eventually add a unified kwarg that replaces vs_module_defs (windows specific) with something that accepts any of:
- libtool flat text file of symbol names
- gnu ld version script
- Apple ld version script
- Windows defs file

And chooses (or converts) the best one.

However, that doesn't exist now.

(I think I'll be in a good position to implement it once some currently pending reworks of the build targets bit of the Meson interpreter go in. So: soon.)

Reply:
fixed